gitlab/resource_gitlab_instance_variable.go

```diff
@@ -0,0 +1,131 @@
+package gitlab
+
+import (
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabInstanceVariable() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabInstanceVariableCreate,
+		Read:   resourceGitlabInstanceVariableRead,
+		Update: resourceGitlabInstanceVariableUpdate,
+		Delete: resourceGitlabInstanceVariableDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"key": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: StringIsGitlabVariableName,
+			},
+			"value": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
+			},
+			"variable_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "env_var",
+				ValidateFunc: StringIsGitlabVariableType,
+			},
+			"protected": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"masked": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+		},
+	}
+}
+
+func resourceGitlabInstanceVariableCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	key := d.Get("key").(string)
+	value := d.Get("value").(string)
+	variableType := stringToVariableType(d.Get("variable_type").(string))
+	protected := d.Get("protected").(bool)
+	masked := d.Get("masked").(bool)
+
+	options := gitlab.CreateInstanceVariableOptions{
+		Key:          &key,
+		Value:        &value,
+		VariableType: variableType,
+		Protected:    &protected,
+		Masked:       &masked,
+	}
+	log.Printf("[DEBUG] create gitlab instance level CI variable %s", key)
+
+	_, _, err := client.InstanceVariables.CreateVariable(&options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(key)
+
+	return resourceGitlabInstanceVariableRead(d, meta)
+}
+
+func resourceGitlabInstanceVariableRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	key := d.Id()
+
+	log.Printf("[DEBUG] read gitlab instance level CI variable %s", key)
+
+	v, _, err := client.InstanceVariables.GetVariable(key)
+	if err != nil {
+		return err
```

Comment:
Can you check if it's a 404 error, and if so use `d.SetId("")`, to handle gracefully if the resource is deleted out-of-band?

Reply:
Done in 10e87f9

Reply:
Thanks, however the xanzy/go-gitlab client we use treats 404 as an error, so the check needs to be inside the err check.

Reply:
Re-done in c379c2d.